src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.Http2.cs

```diff
@@ -370,5 +506,222 @@ public async Task Http2_FlowControl_ClientDoesNotExceedWindows()
                 Assert.Equal(HttpStatusCode.OK, response.StatusCode);
             }
         }
+
+        [ConditionalFact(nameof(SupportsAlpn))]
+        public async Task Http2_InitialWindowSize_ClientDoesNotExceedWindows()
+        {
+            const int DefaultInitialWindowSize = 65535;
+            const int ContentSize = 100_000;
+
+            HttpClientHandler handler = CreateHttpClientHandler();
+            handler.ServerCertificateCustomValidationCallback = TestHelper.AllowAllCertificates;
+            TestHelper.EnsureHttp2Feature(handler);
+
+            var content = new ByteArrayContent(TestHelper.GenerateRandomContent(ContentSize));
+
+            using (var server = Http2LoopbackServer.CreateServer())
+            using (var client = new HttpClient(handler))
+            {
+                Task<HttpResponseMessage> clientTask = client.PostAsync(server.Address, content);
+
+                await server.EstablishConnectionAsync();
+
+                // Bump connection window so it won't block the client.
+                await server.WriteFrameAsync(new WindowUpdateFrame(ContentSize - DefaultInitialWindowSize, 0));
+
+                Frame frame = await server.ReadFrameAsync(TimeSpan.FromSeconds(30));
+                int streamId = frame.StreamId;
+                Assert.Equal(FrameType.Headers, frame.Type);
+                Assert.Equal(FrameFlags.EndHeaders, frame.Flags);
+
+                // Receive up to initial window size
+                int bytesReceived = 0;
+                while (bytesReceived < DefaultInitialWindowSize)
+                {
+                    frame = await server.ReadFrameAsync(TimeSpan.FromSeconds(30));
+                    Assert.Equal(streamId, frame.StreamId);
+                    Assert.Equal(FrameType.Data, frame.Type);
+                    Assert.Equal(FrameFlags.None, frame.Flags);
+                    Assert.True(frame.Length > 0);
+
+                    bytesReceived += frame.Length;
+                }
+
+                Assert.Equal(DefaultInitialWindowSize, bytesReceived);
+
+                // Issue another read. It shouldn't complete yet. Wait a brief period of time to ensure it doesn't complete.
+                Task<Frame> readFrameTask = server.ReadFrameAsync(TimeSpan.FromSeconds(30));
+
+                await Task.Delay(500);
```

Comment:
What's the purpose of all of these Task.Delays?  If it's necessary for correctness, we should instead use synchronization between this and the loopback server to guarantee the needed ordering.  If it's for something else, what is that? :)  And if we must have these, the test should be made Outerloop.

Reply:
It's for correctness. We're trying to ensure that the client isn't sending frames when it's not supposed to. So the problem isn't synchronizing with the server, it's synchronizing with the client, and I don't see any better way to do that... thoughts?

I'm happy to move these to Outerloop.

Reply:
Ah, this is a "make sure that something we expect not to happen doesn't happen" test... yeah, those are notorious for needing timeouts, where in theory you want to do the equivalent of wait forever, but obviously can't, so pick some long-ish timeout that's effectively the same as "forever" for good-enough testing purposes.  Ok.  But yeah, please make it [Outerloop].